planner/core/rule_partition_processor.go

```diff
@@ -114,15 +115,62 @@ func generateHashPartitionExpr(ctx sessionctx.Context, pi *model.PartitionInfo,
 	return exprs[0], nil
 }
 
-func convertToRangeOr(used []int, pi *model.PartitionInfo) partitionRangeOR {
-	if len(used) == 1 && used[0] == FullRange {
-		return fullRange(len(pi.Definitions))
+func (s *partitionProcessor) findUsedPartitions(ctx sessionctx.Context, tbl table.Table, partitionNames []model.CIStr,
+	conds []expression.Expression, columns []*expression.Column, names types.NameSlice) ([]int, []expression.Expression, error) {
+	pi := tbl.Meta().Partition
+	pe, err := generateHashPartitionExpr(ctx, pi, columns, names)
+	if err != nil {
+		return nil, nil, err
 	}
-	ret := make(partitionRangeOR, len(used))
-	for _, i := range used {
-		ret = append(ret, partitionRange{i, i + 1})
+	partIdx := expression.ExtractColumns(pe)
+	colLen := make([]int, 0, len(partIdx))
+	for i := 0; i < len(partIdx); i++ {
+		partIdx[i].Index = i
+		colLen = append(colLen, types.UnspecifiedLength)
 	}
-	return ret
+	datchedResult, err := ranger.DetachCondAndBuildRangeForPartition(ctx, conds, partIdx, colLen)
+	if err != nil {
+		return nil, nil, err
+	}
+	ranges := datchedResult.Ranges
+	used := make([]int, 0, len(ranges))
+	for _, r := range ranges {
+		if r.IsPointNullable(ctx.GetSessionVars().StmtCtx) {
+			if !r.HighVal[0].IsNull() {
+				if len(r.HighVal) != len(partIdx) {
+					used = []int{-1}
```

Comment:
Why here `null` means no partition is pruned?

Reply:
`IsPointNullable` will return true if the range is a point or is a point contain `null`. So if the range is `null` like `[(null, 2), (null, 2)]` it will return true and do pruning.